From 1701140c10e4eb40b2d1e4e4f1c3d7f29d32eeb7 Mon Sep 17 00:00:00 2001
From: Maruthi Srinivas Bayyavarapu <maruthi.srinivas.bayyavarapu@xilinx.com>
Date: Thu, 11 Oct 2018 15:54:26 +0530
Subject: [PATCH 0127/1852] ASoC: xlnx: support multiple sampling rates for SDI

commit 85f59304d42a5f3ccff668157e52cc534687b200 from
https://github.com/Xilinx/linux-xlnx.git

This adds support for SDI playback at various supported sampling rates.

Signed-off-by: Maruthi Srinivas Bayyavarapu <maruthi.srinivas.bayyavarapu@xilinx.com>
Reviewed-by: Sandip Kothari <sandipk@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 sound/soc/xilinx/xlnx_pl_snd_card.c | 29 +++++++++++++++++++++++++++++
 1 file changed, 29 insertions(+)

diff --git a/sound/soc/xilinx/xlnx_pl_snd_card.c b/sound/soc/xilinx/xlnx_pl_snd_card.c
index 0d0411ad83f2..9024db858f9b 100644
--- a/sound/soc/xilinx/xlnx_pl_snd_card.c
+++ b/sound/soc/xilinx/xlnx_pl_snd_card.c
@@ -41,6 +41,30 @@ static struct snd_soc_card xlnx_card = {
 	.owner = THIS_MODULE,
 };
 
+static int xlnx_sdi_card_hw_params(struct snd_pcm_substream *substream,
+				   struct snd_pcm_hw_params *params)
+{
+	struct snd_soc_pcm_runtime *rtd = substream->private_data;
+	struct pl_card_data *prv = snd_soc_card_get_drvdata(rtd->card);
+	u32 sample_rate = params_rate(params);
+
+	switch (sample_rate) {
+	case 32000:
+		prv->mclk_ratio = 576;
+		break;
+	case 44100:
+		prv->mclk_ratio = 418;
+		break;
+	case 48000:
+		prv->mclk_ratio = 384;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
 static int xlnx_hdmi_card_hw_params(struct snd_pcm_substream *substream,
 				    struct snd_pcm_hw_params *params)
 {
@@ -135,6 +159,10 @@ static int xlnx_i2s_card_hw_params(struct snd_pcm_substream *substream,
 	return ret;
 }
 
+static const struct snd_soc_ops xlnx_sdi_card_ops = {
+	.hw_params = xlnx_sdi_card_hw_params,
+};
+
 static const struct snd_soc_ops xlnx_i2s_card_ops = {
 	.hw_params = xlnx_i2s_card_hw_params,
 };
@@ -196,6 +224,7 @@ static struct snd_soc_dai_link xlnx_snd_dai[][XLNX_MAX_PATHS] = {
 		{
 			.name = "xlnx-sdi-playback",
 			SND_SOC_DAILINK_REG(xlnx_sdi_tx),
+			.ops = &xlnx_sdi_card_ops,
 		},
 		{
 			.name = "xlnx-sdi-capture",
-- 
2.31.1

